### PR TITLE
Introducing Applitools and Cypress

### PR DESCRIPTION
**Pull request by megsachdev**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Documentation Update

## Description
Hello!
At times maintaining separate visual test cases becomes a tiring task so using your existing examples setup we tried coming up with something that can ease this process. So I chose to use Cypress as the browser test harness, which Applitools has an SDK for. I started off with writing tests for some of the components so it's sort of a "smoke test" but you can add more if you'd like.

Visual regression involves comparing the pixels visually like a human would, ideally ignoring differences that a human couldn't perceive e.g. 1 pixel changes, differences in how browsers render images, drop shadows, etc. to avoid false positives; but you can still do a strict 1:1 pixel match if that's your preference.

Applitools thankfully provides an SDK that made it trivial to write tests using Cypress. So other than the things I included what you'll need to do is create your own Applitools account and then setup your GitHub CI integration if you want it to run automatically on Pull Requests; optional, but probably a good idea.
Here is the quick reference to Applitools: https://applitools.com/

Every test we write will include selectors for elements. To save a lot of headaches, we should write selectors that are resilient to changes.

Oftentimes we see users run into problems targeting their elements because:
Your application may use dynamic classes or ID’s that change
Your selectors break from development changes to CSS styles or JS behavior

Luckily, it is very easy to avoid both of these problems:
Don’t target elements based on CSS attributes such as: id, class, tag
Don’t target elements that may change their textContent
Add data-* attributes to make it easy to target elements

That's what I have done here, I added data attributes to the existing components to make the code neat and consistent.

To run Applitools:

Create Applitools account
Go to the menu in the top right corner and click the profile icon on the far right.

image

Click on "My API Key" and copy your key

Go to your terminal and paste
``export APPLITOOLS_API_KEY=Your_API_Key_Here``

Go to your directory and run:
``npm cypress run``

In your Applitools dashboard you should see the results of your tests
Screenshot from 2019-08-10 13-33-12

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

**Comment by benhalpern** [comment]
This is really interesting! @jessleenyc @rhymes @lightalloy 

Thoughts or concerns about any of this?

**Comment by jessleenyc** [comment]
Hey @megsachdev! Thanks for the PR. We've heard good things about Cypress and are open to trying it out. However, we don't want to commit to Applitools yet. Would you mind splitting up this PR so it's Cypress only?

In the future, we're more likely to merge full PRs if an issue is created first where we can discuss. Sorry for the request changes! Thanks.

**Comment by rhymes** [comment]
Hi @megsachdev, thanks for your contribution, as @jessleenyc said we'd welcome this PR if Applitools could be removed.

I wanted to add a note:

> That's what I have done here, I added data attributes to the existing components to make the code neat and consistent.

I'm not sure it's the best course, the argument you make about brittle tests is valid but the solution shouldn't be tagging public, production HTML with data attributes that are only needed for end to end testing, the same way we shouldn't write backend code that only aids tests.

I think tests can be brittle if the code they test is badly designed. If an element changes its ID or its class I'd want the test to fail, the chances that I've also restructured the code are high, if the test doesn't fail maybe I'm testing the wrong thing. I understand your point that changing a button from "Welcome to DEV" to "Welcome to DEV!" should not break a test, but it's better than sprinkling the code with data attributes that have no meaning to the app.

Data attributes exist in HTML5 to store custom data, they shouldn't be used to find elements in JS. As with every rule there can be exceptions but this is more a choice than a need. 

BTW this test would fail anyway if the text "Code of Conduct" is changed, so why bother with data attributes at all? :

https://github.com/thepracticaldev/dev.to/pull/3679/files#diff-65289090f597e2b14b99e8c782cb3f68R23-R31

Querying the DOM by data attribute is also quite slow, which eventually would add up time running a suite of tests.

**Comment by benhalpern** [comment]
Hey @megsachdev just to clarify: There seems have been some confusion about whether we had already approved using applitools.

So when we saw this PR, our reaction was "oh hey this is cool, but we're sort of new to applitools and aren't sure if we'd implement it, so let's first merge Cypress". It's not _against_ applitools, just that we are not familiar with it.

That being said, it's possible that we would want to use it, we just need to explore and think it over. With some kind of miscommunication that we'd already taken this step, I hope this isn't too much of a downer. We have had other tools added to the codebase via PR like this, but it's never an easy choice and we don't like to rush into it.

🙂

**Comment by megsachdev** [comment]
Hi @benhalpern 

As discussed I have removed Applitools from the PR and it has just the Cypress test. Request you to please review the same.

File: cypress/integration/visual-testing/homepage-test.spec.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    cy.visit(baseURL);
+    eyesOpen('Home Page');
+  });
+  afterEach(() => {
+    cy.eyesClose();
+  });
+
+  it('should display sign up options', () => {
+    cy.get('#navigation-butt').click();
+    cy.get('#navigation-butt').should('be.visible');
```
Comment (nickytonline) [inline cypress/integration/visual-testing/homepage-test.spec.js]: Cypress' `click` command will only work if the button for the given selector is visible so you don't need to add this check. For reference see https://docs.cypress.io/guides/core-concepts/interacting-with-elements.html#Actionability

File: cypress/integration/visual-testing/homepage-test.spec.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  });
+
+  it('should redirect to code of conduct', () => {
+    const elem = cy.get('[data-cy-href=codeofconduct]');
```
Comment (nickytonline) [inline cypress/integration/visual-testing/homepage-test.spec.js]: It is generally not recommended to set variables values to a cypress command. Consider using an alias instead or chain the commands. https://docs.cypress.io/guides/core-concepts/variables-and-aliases.html

Also, there is no need to use `return`.

```
cy.get('[data-cy-href=codeofconduct]')
	.contains('Code of Conduct')
	.invoke('attr', 'href').then(href => {
      cy.visit(`${baseURL}${href}`);
      cy.wait(2000);
      cy.get('h1')
        .should('be.visible')
        .contains('Code of Conduct');
    });
```

File: cypress/integration/visual-testing/homepage-test.spec.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    elem.contains('Code of Conduct');
+    return elem.invoke('attr', 'href').then(href => {
+      cy.visit(`${baseURL}${href}`);
+      cy.wait(2000);
```
Comment (nickytonline) [inline cypress/integration/visual-testing/homepage-test.spec.js]: In general, it is not recommended to use `wait` unless you have a very good reason as it is an anti-pattern.See https://docs.cypress.io/api/commands/wait.html#Wait-for-an-arbitrary-period-of-milliseconds for more information about this.

File: app/views/articles/_sidebar.html.erb [hunk 4]
```diff
@@ -46,7 +46,7 @@
           <a href="/terms">Terms of Use</a>
           <a href="/contact">Contact</a>
           <a href="/faq">FAQ</a>
-          <a href="/code-of-conduct">Code of Conduct</a>
+          <a href="/code-of-conduct" data-cy-href="codeofconduct">Code of Conduct</a>
```
Comment (nickytonline) [inline app/views/articles/_sidebar.html.erb]: `data-cy` should be adequate enough for e2e tests. It is the recommended data attribute to use by the Cypress team.

File: cypress/integration/visual-testing/homepage-test.spec.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+      cy.get('h1')
+        .should('be.visible')
+        .contains('Code of Conduct');
+    });
+  });
+
+  it('should redirect to about us', () => {
+    cy.get('#loggedoutmenu.menu.logged-out').invoke('show');
```
Comment (nickytonline) [inline cypress/integration/visual-testing/homepage-test.spec.js]: Consider using the `data-cy` attribute instead. Relying on selectors that are CSS classes or IDs may change over time causing an e2e test fail.

**Comment by cmwhited** [comment]
@megsachdev can you take a look at the comment above as well as the suggestions?

**Comment by megsachdev** [comment]
Hi @nickytonline 
Thanks for the review, I have resolved all the issues according to the comments.

**Comment by cmwhited** [comment]
@nickytonline have you had a chance to look at this since the last update?

**Comment by nickytonline** [comment]
@megsachdev, thanks again for the PR. 😄 I see you made the changes.

Having said that, there's been some discussion in #382 about potentially using another e2e testing framework, specifically Test Café. I'm not on the DEV team, but we should probably wait on this PR until the DEV team has decided on what makes the most sense for E2E testing.

**Comment by benhalpern** [comment]
Hey, we haven't made any progress on deciding whether or not we can merge this.

Very sorry about the confusion around this one. I'm going to close this for now because we don't have any plans on the near horizon to pick back up this conversation and we don't want to keep this in limbo.

**Pull request closed without merging**